File: chan/operatorrr.cc
#include "chan.ih"

template<typename T>
chan<T> &chan<T>::operator>>(chan<T> ch, T &val) {
    val = this->d_impl->receive();
    return ch;
}

template<typename T>
chan<T> &chan<T>::operator>>(chan<T> ch, T &&) {
    this->d_impl->receive();
    return ch;
}

File: chan/operatorrr.f
template<typename T>
chan<T> &chan<T>::operator>>(T &val) {
    this->d_impl->receive(val);
    return *this;
}

template<typename T>
chan<T> &chan<T>::operator>>(T &&) {
    this->d_impl->receive();
    return *this;
}